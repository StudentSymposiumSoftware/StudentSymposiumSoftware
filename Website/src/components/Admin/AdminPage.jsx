import "./AdminPage.css"

// Firebase Imports
import { storage, database, authentication } from "../../firebase/firebaseConfig";
import { ref as dbRef, onValue } from "firebase/database";
import { uploadBytes, ref as storageRef } from "firebase/storage";
import { signInWithRedirect, GoogleAuthProvider } from "firebase/auth";

import { useAuthState } from "react-firebase-hooks/auth"

// React Imports
import { useState, useEffect } from "react";

/* v8 ignore start */
export function AuthAdmin() {
    function uploadFile(file) {
        var year = document.getElementById("year").value;
        year = year == ""? new Date().getFullYear() : year;
        if (file == null) {
            alert("Please select a file to upload.");
            return;
        }
        const fileRef = storageRef(storage, `symposium/${year}/${year}-upload.csv`);
        uploadBytes(fileRef, file).then(() => {
            alert("File uploaded successfully!");
        }).catch((error) => {
            console.error("Error uploading file:", error);
            alert("Error uploading file. Please try again.");
        });
    }
    
    return (
        <div className="text-center flex flex-col items-center">
            <div className="mb-[10px]">
                <div className="font-bold text-[26px]">Symposium Year End File Upload Utility</div>
                <div className="font-bold text-[18px]">Please Upload The Excel File Here</div>
            </div>
            <form className="flex flex-col items-center">
                <input type="text" id="year" name="year" placeholder={"Year (Default: "+ new Date().getFullYear()+")"} className="block p-2 bg-white text-black rounded-md mb-5"/>
                <input type="file" id="file" name="file" accept=".csv" className="hidden" onChange={(e) => uploadFile(e.target.files[0])}/>
                <label for="file" className="button mb-2 text-[20px] cursor-pointer bg-white text-black rounded-md p-2">Upload .csv Symposium File</label>
            </form>
        </div>
    )
}

export function Loading() {  
    return (
        <div>
            <h1>Loading...</h1>
        </div>
    )
}

export function AdminPage() {
    const [authUser, authLoading] = useAuthState(authentication) // auth user object (used to obtain other user object)
    const [page, setPage] = useState("loading");
    useEffect(() => {
        if (authUser && authLoading === false) {
            onValue(dbRef(database, `users/${authUser.uid}`), (userData) => {
                userData = userData.val();
                if (userData != null && userData.authorized) {
                    setPage("admin")
                } else {
                    window.location.href = "/";
                }
            });
        } else if (authLoading === false) {
            signInWithRedirect(authentication, new GoogleAuthProvider()).then((result) => {
                console.log("Signed in as:", result.user.displayName);
            }).catch((error) => {
                console.error("Error signing in:", error);
            }
            );
            setPage("loading");
        }
    }, [authLoading]);
    return (
        <main>
            {page === "admin" && <AuthAdmin />}
            {page === "loading" && <Loading />}
        </main>
    )
}

export default AdminPage;
/* v8 ignore stop */