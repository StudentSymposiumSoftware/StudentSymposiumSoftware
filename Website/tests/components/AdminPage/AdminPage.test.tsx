import React from 'react';
import { render } from "@testing-library/react";
import { expect, it } from "vitest";

import { AuthAdmin, Loading, AdminPage } from '../../../src/components/Admin/AdminPage';

it('Admin Page renders correctly', () => {
    const tree = render(<AdminPage/>)
    expect(tree).toMatchSnapshot();
});

it('Loading Module Renders Correctly', () => {
    const tree = render(<Loading/>)
    expect(tree).toMatchSnapshot();
});

it('AuthAdmin Module Renders Correctly', () => {
    const tree = render(<AuthAdmin/>)
    expect(tree).toMatchSnapshot();
});