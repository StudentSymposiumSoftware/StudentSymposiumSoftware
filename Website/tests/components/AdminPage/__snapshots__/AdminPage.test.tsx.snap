// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Admin Page renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main>
        <div>
          <h1>
            Loading...
          </h1>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main>
      <div>
        <h1>
          Loading...
        </h1>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AuthAdmin Module Renders Correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main>
        <div>
          <h1>
            Loading...
          </h1>
        </div>
      </main>
    </div>
    <div>
      <div>
        <h1>
          Loading...
        </h1>
      </div>
    </div>
    <div>
      <div
        class="text-center flex flex-col items-center"
      >
        <div
          class="mb-[10px]"
        >
          <div
            class="font-bold text-[26px]"
          >
            Symposium Year End File Upload Utility
          </div>
          <div
            class="font-bold text-[18px]"
          >
            Please Upload The Excel File Here
          </div>
        </div>
        <form
          class="flex flex-col items-center"
        >
          <input
            class="block p-2 bg-white text-black rounded-md mb-5"
            id="year"
            name="year"
            placeholder="Year (Default: 2025)"
            type="text"
          />
          <input
            accept=".csv"
            class="hidden"
            id="file"
            name="file"
            type="file"
          />
          <label
            class="button mb-2 text-[20px] cursor-pointer bg-white text-black rounded-md p-2"
            for="file"
          >
            Upload .csv Symposium File
          </label>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="text-center flex flex-col items-center"
    >
      <div
        class="mb-[10px]"
      >
        <div
          class="font-bold text-[26px]"
        >
          Symposium Year End File Upload Utility
        </div>
        <div
          class="font-bold text-[18px]"
        >
          Please Upload The Excel File Here
        </div>
      </div>
      <form
        class="flex flex-col items-center"
      >
        <input
          class="block p-2 bg-white text-black rounded-md mb-5"
          id="year"
          name="year"
          placeholder="Year (Default: 2025)"
          type="text"
        />
        <input
          accept=".csv"
          class="hidden"
          id="file"
          name="file"
          type="file"
        />
        <label
          class="button mb-2 text-[20px] cursor-pointer bg-white text-black rounded-md p-2"
          for="file"
        >
          Upload .csv Symposium File
        </label>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Loading Module Renders Correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main>
        <div>
          <h1>
            Loading...
          </h1>
        </div>
      </main>
    </div>
    <div>
      <div>
        <h1>
          Loading...
        </h1>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        Loading...
      </h1>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
